### PR TITLE
Support add to PATH on macOS

### PR DESCRIPTION
**Pull request by ryanccn**
~~Incredibly early support. Permissions system is a wreck, a [future Qt API](https://codereview.qt-project.org/c/qt/qtbase/+/395084) is still under review, and Full Disk Access has been 
confirmed to not work.~~

~~I'll try and find how VS Code does it, since Electron is about as wack as Qt is :/~~

Fixed using an approach inspired from VSCode! This one uses `osascript` to run a piece of AppleScript that asks for administrator privileges to run a command. Installs a `prismlauncher` symlink to the application binary :)

### CODE REVIEW

**Comment by ghost** [comment]
can i know the reason for need of a path for prism?

**Comment by ghost** [comment]
Could I know the use case for adding Prism path?
This PR does not make any reference to any issues.

**Comment by ryanccn** [comment]
> Could I know the use case for adding Prism path? This PR does not make any reference to any issues.

@mistic316 Running Prism from the command line. Normally you'd do `/Applications/PrismLauncher.app/Contents/MacOS/prismlauncher` to execute it, this PR just ~~is supposed to~~ adds a convenient button to make a symlink in `/usr/local/bin` to enable you to directly access with `prismlauncher`.

**Comment by ghost** [comment]
@ryanccn 
For this, a user needs to give administrative permission to Prism to add a path (since this is modifying a system directory), from my understanding.
For a simple minecraft launcher, this kind of privilege permission is unnecessary (think of it running as root).

Again, in what scenario, running prism from terminal would benefit? What is the use case? What issues is this trying to fix?

**Comment by Scrumplex** [comment]
Shouldn't this be handled by package managers?

**Comment by ryanccn** [comment]
What no on macOS there's no install process at all

Basically you drag the .app to the Applications folder but it doesn't add it to PATH unless you make a symlink to the app from, for instance, `/usr/local/bin/prismlauncher`

It's not standard practice for apps to be in the PATH on macOS unless there's a special reason to do so (here, the CLI launcher)

**Comment by DioEgizio** [comment]
> Shouldn't this be handled by package managers?

most people are just gonna just drag and drop to /Applications or ~/Applications

but i do agree that this is quite hacky and unnecessary

**Comment by Scrumplex** [comment]
I thought about brew/macports

**Comment by ryanccn** [comment]
I don't know about ports, but usually you do not symlink casks, the ones that do install to PATH (such as 1Password's `op`) use .pkg package installers.

This is useful for when you want to be able to run `prismlauncher` anywhere and have it registered in the PATH, since if you wanted ease of access you would've made the symlink manually anyway. VS Code has this CLI feature because its CLI can be used to open files / folders in VS Code, and launching instances is a similar use case IMO

**Comment by ryanccn** [comment]
review wen 💀

**Comment by AshtakaOOf** [comment]
Never duh
edit: @ryanccn It's a joke :skull:

**Comment by Scrumplex** [comment]
Can't users just create the symlink manually?

**Comment by ryanccn** [comment]
well it's simpler to have a button that does this in the launcher

**Comment by Scrumplex** [comment]
Well, it makes sense for an IDE like VS Code to have that, as a user is going to start it frequently from the terminal. But for a Minecraft launcher, I don't really see a reason why you would need that. If you are the kind of person who would frequently run a launcher from the terminal, it would be better if you just did it yourself.

**Comment by ryanccn** [comment]
Well, there's nothing to lose if you do add this feature and I've talked to some macOS users who would love to see this QoL feature even if it's a very simple one.

Other than that, this also simplifies launching instances from, for instance, Shortcuts or Raycast Script Commands, for even non technical users, because you only have to write `prismlauncher` in the "Run shell command" action instead of having to go into the .app and copy the whole path to the binary.

**Comment by ghost** [comment]
@ryanccn 
> This one uses osascript to run a piece of AppleScript that asks for administrator privileges to run a command.

I still think this feature is not useful. Specially giving simple minecraft launcher admin privilege permission in order to use it.
Still, if you are going to add it, please make it something toggleable like in VS Code that is **not** enabled by default.

**Comment by ryanccn** [comment]
> @ryanccn
> 
> > This one uses osascript to run a piece of AppleScript that asks for administrator privileges to run a command.
> 
> I still think this feature is not useful. Specially giving simple minecraft launcher admin privilege permission in order to use it. Still, if you are going to add it, please make it something toggleable like in VS Code that is **not** enabled by default.

It isn't enabled by default, have you even tested the feature I'm proposing here

**Review by flowln — APPROVED** [review]
i'm fine with this change

just please de-indent the pre-processor things (`#ifdef`and `#endif`)

File: launcher/ui/MainWindow.cpp [hunk 1]
```diff
@@ -1929,6 +1948,21 @@ void MainWindow::on_actionClearMetadata_triggered()
...
+
+    qDebug() << "Symlinking" << binaryPath << "to /usr/local/bin/prism";
+    auto outcome = QProcess::execute("/usr/bin/osascript", QStringList()<< "-e" << tr("do shell script \"mkdir -p /usr/local/bin && ln -sf '%1' '/usr/local/bin/prismlauncher'\" with administrator privileges").arg(binaryPath));
+
+    if (!outcome) {
+        QMessageBox::information(this, tr("Added Prism to PATH"), tr("Prism was successfully added to your PATH. You can now run it with `prismlauncher` in your Terminal. Enjoy!"));
+    } else {
+        QMessageBox::critical(this, tr("Failed to add Prism to PATH"), tr("Failed to add Prism to PATH :("));
+    }
+}
```
Comment (Scrumplex) [inline launcher/ui/MainWindow.cpp]:
```suggestion
void MainWindow::on_actionAddToPATH_triggered() {
    auto binaryPath = APPLICATION->applicationFilePath();
    auto targetPath = QString("/usr/local/bin/%1").arg(BuildConfig.LAUNCHER_APP_BINARY_NAME);

    qDebug() << "Symlinking" << binaryPath << "to" << targetPath;
    
    QStringList args;
    args << "-e";
    args << QString("do shell script \"mkdir -p /usr/local/bin && ln -sf '%1' '%2'\" with administrator privileges").arg(binaryPath, targetPath);
    auto outcome = QProcess::execute("/usr/bin/osascript", args);

    if (!outcome) {
        QMessageBox::information(this, tr("Successfully added %1 to PATH").arg(BuildConfig.LAUNCHER_DISPLAYNAME), tr("%1 was successfully added to your PATH. You can now start it by running `%2`.").arg(BuildConfig.LAUNCHER_DISPLAYNAME, BuildConfig.LAUNCHER_APP_BINARY_NAME));
    } else {
        QMessageBox::critical(this, tr("Failed to add %1 to PATH").arg(BuildConfig.LAUNCHER_DISPLAYNAME), tr("An error occurred while trying to add %1 to PATH").arg(BuildConfig.LAUNCHER_DISPLAYNAME));
    }
}
```

Comment (Scrumplex) [reply launcher/ui/MainWindow.cpp]: ^ needs clang-format

**Review by Scrumplex — CHANGES_REQUESTED** [review]

**Review by Scrumplex — APPROVED** [review]

**Comment by Scrumplex** [comment]
Someone needs to test this on macOS please (Don't have the means rn to verify)

**Pull request merged**